cloud sdk 404 errors causing cache to be disabled

Every handful of builds I am seeing issues that look like this:

`The remote build cache was disabled during the build due to errors.`
with a trimmed stack that is roughly:
```
org.gradle.caching.BuildCacheException: Unable to load 'cache_key' from Google Cloud Storage bucket 'bucket_name'.
at net.idlestate.gradle.caching.GCSBuildCacheService.load(GCSBuildCacheService.kt:86)

Caused by: com.google.cloud.storage.StorageException: 404 Not Found
No such object: bucket_name/cache_key
at com.google.cloud.storage.spi.v1.HttpStorageRpc.translate(HttpStorageRpc.java:226)

Caused by: com.google.api.client.googleapis.json.GoogleJsonResponseException: 404 Not Found
```

Given that a 404 is a simple not found that seems it may be an eventual consistency type issue, should we special case the catch to check for that 404 and simple return false rather than re-throwing the BuildCacheException?

--- a/README.adoc
+++ b/README.adoc
@@ -37,7 +37,7 @@ buildscript {
     }
 
     dependencies {
-        classpath 'net.idlestate:gradle-gcs-build-cache:1.0.0'
+        classpath 'net.idlestate:gradle-gcs-build-cache:1.0.1'
     }
 }
 
@@ -71,7 +71,7 @@ buildscript {
     }
 
     dependencies {
-        classpath 'net.idlestate:gradle-gcs-build-cache:1.0.0'
+        classpath 'net.idlestate:gradle-gcs-build-cache:1.0.1'
     }
 }
 
@@ -108,7 +108,7 @@ initscript {
     }
 
     dependencies {
-        classpath 'net.idlestate:gradle-gcs-build-cache:1.0.0'
+        classpath 'net.idlestate:gradle-gcs-build-cache:1.0.1'
     }
 }
 

--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -3,23 +3,23 @@ import org.jlleitschuh.gradle.ktlint.reporter.ReporterType
 
 plugins {
     `java-gradle-plugin`
-    kotlin("jvm") version "1.3.20"
-    id("com.github.hierynomus.license") version "0.14.0"
+    kotlin("jvm") version "1.3.31"
+    id("com.github.hierynomus.license") version "0.15.0"
     id("com.gradle.plugin-publish") version "0.10.1"
+    `kotlin-dsl`
     `maven-publish`
-    id("org.jlleitschuh.gradle.ktlint") version "7.0.0"
+    id("org.jlleitschuh.gradle.ktlint") version "8.2.0"
 }
 
 group = "net.idlestate"
-version = "1.0.0"
+version = "1.0.1"
 
 repositories {
-    mavenCentral()
     jcenter()
 }
 
 dependencies {
-    implementation("com.google.cloud:google-cloud-storage:1.61.0")
+    implementation("com.google.cloud:google-cloud-storage:1.86.0")
     implementation(kotlin("stdlib-jdk8"))
 }
 
@@ -44,12 +44,8 @@ pluginBundle {
     tags = listOf("build-cache", "gcs", "Google Cloud Storage", "cache")
 }
 
-val compileKotlin: KotlinCompile by tasks
-compileKotlin.kotlinOptions {
-    jvmTarget = JavaVersion.VERSION_1_8.toString()
-}
-
-val compileTestKotlin: KotlinCompile by tasks
-compileTestKotlin.kotlinOptions {
-    jvmTarget = JavaVersion.VERSION_1_8.toString()
+tasks.withType<KotlinCompile>().all {
+    kotlinOptions {
+        jvmTarget = JavaVersion.VERSION_1_8.toString()
+    }
 }

--- a/gradle/wrapper/gradle-wrapper.properties
+++ b/gradle/wrapper/gradle-wrapper.properties
@@ -1,5 +1,5 @@
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-5.1.1-bin.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-5.5.1-all.zip
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists

--- a/src/main/kotlin/net/idlestate/gradle/caching/GCSBuildCache.kt
+++ b/src/main/kotlin/net/idlestate/gradle/caching/GCSBuildCache.kt
@@ -22,5 +22,8 @@ import org.gradle.caching.configuration.AbstractBuildCache
  *
  * @author Thorsten Ehlers (thorsten.ehlers@googlemail.com) (initial creation)
  */
-open class GCSBuildCache(var credentials: String? = "", var bucket: String? = "", var refreshAfterSeconds: Int? = 0)
-    : AbstractBuildCache()
+abstract class GCSBuildCache(
+    var credentials: String? = "",
+    var bucket: String? = "",
+    var refreshAfterSeconds: Int? = 0
+) : AbstractBuildCache()

--- a/src/main/kotlin/net/idlestate/gradle/caching/GCSBuildCachePlugin.kt
+++ b/src/main/kotlin/net/idlestate/gradle/caching/GCSBuildCachePlugin.kt
@@ -25,10 +25,11 @@ import org.gradle.api.initialization.Settings
  */
 class GCSBuildCachePlugin : Plugin<Settings> {
     override fun apply(settings: Settings) {
-        val buildCacheConfiguration = settings.buildCache
-        buildCacheConfiguration.registerBuildCacheService(GCSBuildCache::class.java, GCSBuildCacheServiceFactory::class.java)
-
-        val cache = buildCacheConfiguration.remote(GCSBuildCache::class.java)
-        cache.isPush = true
+        settings.buildCache {
+            registerBuildCacheService(GCSBuildCache::class.java, GCSBuildCacheServiceFactory::class.java)
+            remote(GCSBuildCache::class.java) {
+                isPush = true
+            }
+        }
     }
 }

--- a/src/main/kotlin/net/idlestate/gradle/caching/GCSBuildCacheService.kt
+++ b/src/main/kotlin/net/idlestate/gradle/caching/GCSBuildCacheService.kt
@@ -43,7 +43,11 @@ class GCSBuildCacheService(credentials: String, val bucketName: String, val refr
 
     init {
         try {
-            val storage = StorageOptions.newBuilder().setCredentials(ServiceAccountCredentials.fromStream(FileInputStream(credentials))).build().service
+            val storage = StorageOptions.newBuilder()
+                .setCredentials(ServiceAccountCredentials.fromStream(FileInputStream(credentials)))
+                .build()
+                .service
+
             bucket = storage.get(bucketName) ?: throw BuildCacheException("$bucketName is unavailable")
         } catch (e: FileNotFoundException) {
             throw BuildCacheException("Unable to load credentials from $credentials.", e)
@@ -83,7 +87,12 @@ class GCSBuildCacheService(credentials: String, val bucketName: String, val refr
                 return true
             }
         } catch (e: StorageException) {
-            throw BuildCacheException("Unable to load '${key.hashCode}' from Google Cloud Storage bucket '$bucketName'.", e)
+            // https://github.com/googleapis/google-cloud-java/issues/3402
+            if (e.message?.contains("404") == true) {
+                return false
+            } else {
+                throw BuildCacheException("Unable to load '${key.hashCode}' from Google Cloud Storage bucket '$bucketName'.", e)
+            }
         }
 
         return false

